enterprise/frontend/src/embedding-sdk/lib/interactive-question/run-question-on-query-change.ts

```diff
@@ -0,0 +1,89 @@
+import _ from "underscore";
+
+import type { SdkQuestionResult } from "embedding-sdk/types/question";
+import { loadMetadataForCard } from "metabase/questions/actions";
+import { getMetadata } from "metabase/selectors/metadata";
+import { syncVizSettingsWithSeries } from "metabase/visualizations/lib/sync-settings";
+import * as Lib from "metabase-lib";
+import Question from "metabase-lib/v1/Question";
+import type { Dispatch, GetState } from "metabase-types/store";
+
+import { runQuestionQuerySdk } from "./run-question-query";
+
+interface RunQuestionOnQueryChangeParams {
+  previousQuestion: Question;
+  nextQuestion: Question;
+  originalQuestion?: Question;
+  shouldStartAdHocQuestion?: boolean;
+  queryResults?: any[];
+}
+
+export const runQuestionOnQueryChangeSdk =
+  (params: RunQuestionOnQueryChangeParams) =>
+  async (
+    dispatch: Dispatch,
+    getState: GetState,
+  ): Promise<SdkQuestionResult> => {
+    let {
+      previousQuestion,
+      nextQuestion,
+      originalQuestion,
+      queryResults,
+      shouldStartAdHocQuestion = true,
+    } = params;
+
+    const { isEditable } = Lib.queryDisplayInfo(nextQuestion.query());
+
+    const shouldTurnIntoAdHoc =
+      shouldStartAdHocQuestion && nextQuestion.isSaved() && isEditable;
+
+    if (shouldTurnIntoAdHoc) {
+      nextQuestion = nextQuestion.withoutNameAndId();
+
+      // When the dataset query changes, we should change the question type,
+      // to start building a new ad-hoc question based on a dataset
+      // NOTE: we do not support model and metric questions in the SDK yet.
+      if (nextQuestion.type() === "model" || nextQuestion.type() === "metric") {
+        nextQuestion = nextQuestion.setType("question");
+      }
+    }
+
+    if (queryResults) {
```

Comment:
Can we maybe create a shared function/action that does all of this? Syncs viz settings, loads metadata, changes to adhoc

Reply:
Good idea! Will do.